src/cache/disk.rs

```diff
@@ -18,17 +18,59 @@ use futures_03::executor::ThreadPool;
 use lru_disk_cache::Error as LruError;
 use lru_disk_cache::LruDiskCache;
 use std::ffi::OsStr;
+use std::mem;
 use std::path::{Path, PathBuf};
 use std::sync::{Arc, Mutex};
+use std::thread::{self, JoinHandle};
 use std::time::{Duration, Instant};
 
 use crate::errors::*;
 
+enum DiskCacheInit {
+    Initializing(JoinHandle<LruDiskCache>),
+    Joining,
+    Initialized(Arc<Mutex<LruDiskCache>>),
+}
+
+impl DiskCacheInit {
+    fn new(root: &OsStr, max_size: u64) -> Self {
+        let root = root.to_os_string();
+        let worker = thread::spawn(move || {
+            LruDiskCache::new(root, max_size).expect("Couldn't instantiate disk cache!")
+        });
+
+        Self::Initializing(worker)
+    }
+
+    fn opt_lru(&self) -> Option<&Arc<Mutex<LruDiskCache>>> {
+        if let Self::Initialized(lru) = self {
+            Some(lru)
+        } else {
+            None
+        }
+    }
+
+    fn lru(&mut self) -> &Arc<Mutex<LruDiskCache>> {
+        let state = mem::replace(self, DiskCacheInit::Joining);
+        if let Self::Initializing(worker) = state {
+            let lru = worker.join().unwrap();
+            *self = DiskCacheInit::Initialized(Arc::new(Mutex::new(lru)));
+        } else {
+            *self = state;
+        }
+
+        if let Self::Initialized(lru) = self {
+            lru
+        } else {
+            unreachable!("The disk cache should have been initialized")
+        }
+    }
+}
+
 /// A cache that stores entries at local disk paths.
-#[derive(Clone)]
 pub struct DiskCache {
     /// `LruDiskCache` does all the real work here.
-    lru: Arc<Mutex<LruDiskCache>>,
+    lru_init: Mutex<DiskCacheInit>,
```

Comment:
Something like `LazyInit` would be best, but it initializes on the same thread. Is there something that I'm not aware of like that in the crate ecosystem already?